File: DAA/Quick Sort/quickSort.cpp
#include<iostream>
using namespace std;

void swap(int i, int j, int arr[]){
    int temp;
    temp = arr[j];
    arr[j]=arr[i];
    arr[i]=temp;
}

int partition(int low, int high, int arr[]){
    int pivot = arr[low];
    int i=low;
    int j=high;
    while(i < j){
        while(arr[i] <= pivot){
            i++;
        }
        while(arr[j] > pivot){
            j--; // imp see da sign bruh
        }
        if(i<j){
            swap(i,j,arr);
        }
    }
    swap(j,low,arr);
    return j;
}

void quickSort(int low, int high, int arr[]){

    if(high > low){
        int p = partition(low, high, arr);
        quickSort(low,p-1,arr);
        quickSort(p+1,high,arr);
    }
}

void dispArray(int lenArr, int arr[]){
    for(int i=1; i<=lenArr; i++){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
}

int main(){
    
    int lenArr;
    cout<<"Enter Array Length: ";
    cin>>lenArr;

    int *arr = new int[lenArr+1];

    
    for(int i=1; i<=lenArr; i++){
        cout<<"Enter values: ";
        cin>>arr[i];
    }

    cout<<endl;
    cout<<"Array Before: ";
    dispArray(lenArr,arr);

    quickSort(0,lenArr,arr);

    cout<<"Array After: ";
    dispArray(lenArr,arr);
}

File: DAA/Quick Sort/QuickSort.java
public class QuickSort {

    static void quickSort(int arr[], int i, int j){
        if(i<j){
            int p = partition(arr, i, j);
            quickSort(arr, i, p-1);
            quickSort(arr, p+1, j);
        }
    }

    static int partition(int arr[],int q, int r){
        int pivot=arr[r];
        
        int m=(q-1);

        for(int n=q;n<r;n++){
            if(arr[n] < pivot){
                m++;
                swap(arr, m, n);
            }
        }
        swap(arr, m+1, r);
        return m+1;
    }

    static void swap(int arr[], int s, int t){
        int temp;
        temp = arr[s];
        arr[s] = arr[t];
        arr[t] = temp;
    }

    public static void main(String[] args){
        int[] arr = {2,8,7,1,3,5,6,4};
        quickSort(arr,0,arr.length-1);

        for(int x = 0; x<arr.length; x++){
            System.out.print(arr[x] + " ");
        }
        System.out.println(" ");
     }

}
